File: python/binary_search.py
"""
Binary search in a sorted list
https://en.wikipedia.org/wiki/Binary_search_algorithm
"""
from bisect import bisect_left


def pythonic_binary_search(lis, elem):
    """
    Return the left most index of matched element in lis with elem using bisect_left
    Return -1 if match not found
    :param lis: sorted list
    :param elem: element to search
    """
    index = bisect_left(lis, elem)
    if index != len(lis) and lis[index] == elem:
        return index
    return -1


def binary_search(lis, elem, start, end):
    """
    Return index of matched position of elem in lis between index start and end (inclusive) or return -1
    :param lis: Sorted list
    :param elem: element to search
    """
    assert 0 <= start <= end < len(lis)
    if end == start:
        # If there is only one element to search, check that element only
        if lis[start] == elem:
            return start
        return -1

    mid = (start + end) // 2
    mid_elem = lis[mid]
    if mid_elem == elem:
        return mid
    if mid_elem < elem:
        # If middle element is less than our required element then search in the right of array
        if end > mid:
            return binary_search(lis, elem, mid + 1, end)
        return -1
    if mid_elem > elem:
        # If middle element is larger than our required element then search in the left of array
        if start < mid:
            return binary_search(lis, elem, start, mid - 1)
        return -1


def main():
    lis1 = [1, 2, 3, 4, 6, 7, 9]
    assert binary_search(lis1, 5, 0, 5) == -1
    assert pythonic_binary_search(lis1, 5) == -1

    lis2 = [-1, 0, 1, 5, 6, 9, 15, 15]
    assert pythonic_binary_search(lis2, 15) == 6
    assert binary_search(lis2, 15, 0, 7) != -1  # Our custom binary search can't tell the left most index

    lis3 = [-1, 0, 0, 0, 0, 1]
    assert pythonic_binary_search(lis3, 0) == 1
    assert binary_search(lis3, 0, 0, 5) != -1


if __name__ == '__main__':
    main()

File: erlang/binary_search.erl
-include_lib("eunit/include/eunit.hrl").

-module(binary_search).
-export([start/0]).

% Binary search implementation in Erlang
% https://en.wikipedia.org/wiki/Binary_search_algorithm


% Entrypoint that calls the function with an additional accumulator
% to save the index of the found element
binary_search(Haystack, Needle) ->
  binary_search(Haystack, Needle, 0).

binary_search(Haystack, Needle, PrevIndex) ->
  % Get the middle element
  HaystackLength = length(Haystack),
  Middle = (HaystackLength + 1) div 2,
  case Middle of
    % If there is no middle, the list is empty, so no element is found
    0 -> -1;
    _ ->
      % In any other case, get the element at the middle and compare it
      % with the element looked for
      Item = lists:nth(Middle,Haystack),
      case Item of
        % If it is the element looked for,
        % We return the sum of indexes and substract 1
        % Since it is a zero-based index search
        Item when Item =:= Needle ->
          PrevIndex + Middle - 1;
        % When the item is either greater or smaller than the
        % value looked for, search in the left or right lists
        % and also send the current index of the search
        Item when Item > Needle ->
          Left = lists:sublist(Haystack, HaystackLength - Middle),
          binary_search(Left, Needle, PrevIndex + Middle);
        Item when Item < Needle ->
          Right = lists:nthtail(Middle, Haystack),
          binary_search(Right, Needle, PrevIndex + Middle)
        end
  end.

start() ->
  Test1 = [1, 2, 3, 4, 6, 7, 9],
  Test2 = [-1, 0, 1, 5, 6, 9, 15, 15],
  Test3 = [-1, 0, 0, 0, 0, 1],
  Test4 = [],

  ?assert(binary_search(Test1,5) =:= -1),
  ?assert(binary_search(Test2,15) =:= 6),

  % Finding the left-most occurrence is not implemented
  % but it would find the element in the first iteration
  ?assert(binary_search(Test3,0) =/= -1),

  ?assert(binary_search(Test4, 10) =:= -1).
